File: scripts/microbe_stage/spawn_system.lua

--------------------------------------------------------------------------------
-- SpawnedComponent
--
-- Holds information about an entity spawned by spawnComponent
--------------------------------------------------------------------------------
SpawnedComponent = class(
    function(self)

        self.spawnRadiusSqr = 1000
    end
)

SpawnedComponent.TYPE_NAME = "SpawnedComponent"

SPAWN_INTERVAL = 100 --Time between spawn cycles

function SpawnedComponent:load(storage)

    self.spawnRadiusSqr = storage:get("spawnRadius", 1000)
end


function SpawnedComponent:storage(storage)

    storage:set("spawnRadius", self.spawnRadiusSqr)

end

REGISTER_COMPONENT("SpawnedComponent", SpawnedComponent)


--------------------------------------------------------------------------------
-- SpawnSystem
--
-- System for spawning and despawning entities
--------------------------------------------------------------------------------
SpawnSystem = class(
    LuaSystem,
    function(self)

        LuaSystem.create(self)
        
        self.entities = EntityFilter.new(
            {
                SpawnedComponent
            }
        )
        self.nextId = 1
        self.spawnTypes = {} --Keeps track of factory functions.
        
        self.playerPosPrev = nil --A Vector3 that remembers the player's position in the last spawn cycle
        
        self.timeSinceLastCycle = 0 --Stores how much time has passed since the last spawn cycle
        
        currentSpawnSystem = self
        
    end
)

currentSpawnSystem = nil

-- Override from System
function SpawnSystem:init(gameState)
    LuaSystem.init(self, "SpawnSystem", gameState)
    self.entities:init(gameState.wrapper)
end

-- Override from System
function SpawnSystem:shutdown()
    self.entities:shutdown()
    LuaSystem.shutdown(self)
end

-- Adds a new type of entity to spawn in the SpawnSystem
-- Also returns an unique id to reference it.
-- Note that SpawnedComponent will be added automatically and should not be done by the factory function
--
-- @param factoryFunction
--  The function called by the SpawnSystem to create the entity. It should have two
--  parameters, x and y positions, and it should return the new entity.
--
-- @param spawnDensity
--  On average, the number of entities of the given type per square unit.
--
-- @param spawnRadius
--  The distance from the player that the entity can spawn or despawn.
function SpawnSystem:addSpawnType(factoryFunction, spawnDensity, spawnRadius)
    newSpawnType = {}
    newSpawnType.factoryFunction = factoryFunction
    newSpawnType.spawnRadius = spawnRadius
    newSpawnType.spawnRadiusSqr = spawnRadius * spawnRadius
    
    --spawnFrequency is on average how many entities should pass the first condition
    --in each spawn cycle (See _doSpawnCycle). spawnRadius^2 * 4 is used because
    --that is the area of the square region where entities attempt to spawn.
    newSpawnType.spawnFrequency = spawnDensity * spawnRadius * spawnRadius * 4
    
    id = "spawnType" .. self.nextId
    self.nextId = self.nextId + 1
    self.spawnTypes[id] = newSpawnType
    return id
end

--removes an element
--
-- @param id
--  The id of the element to be removed
function SpawnSystem:removeSpawnType(id)
    self.spawnTypes[id] = nil
end

-- For each entity type, spawns the appropriate number of entities within the spawn
-- radius of the player at its current location but outside of the spawn radius of the player
-- at its location during the previous spawn cycle.
function SpawnSystem:_doSpawnCycle()    
    local player = Entity.new(PLAYER_NAME, self.gameState.wrapper)
    
    local playerNode = getComponent(player, OgreSceneNodeComponent)
    local playerPos = playerNode.transform.position
    
    --Initialize previous player position if necessary
    if self.playerPosPrev == nil then
        self.playerPosPrev = Vector3(playerPos.x, playerPos.y, playerPos.z)
    end
    
    for _, entity in pairs(self.entities:removedEntities()) do
        self.microbes[entityId] = nil
    end
    for _, entityId in pairs(self.entities:addedEntities()) do
        local microbe = Microbe.new(Entity.new(entityId, self.gameState.wrapper))
        self.microbes[entityId] = microbe
    end
    self.entities:clearChanges()
    
    --Despawn entities    
    for _, entityId in pairs(self.entities:entities()) do
        entity = Entity.new(entityId, self.gameState.wrapper)
        local spawnComponent = getComponent(entity, SpawnedComponent)
        local sceneNode = getComponent(entity, OgreSceneNodeComponent)
        local entityPos = sceneNode.transform.position
        local distSqr = playerPos:squaredDistance(entityPos)
        
        --Destroy and forget entities outside the spawn radius.
        if distSqr >= spawnComponent.spawnRadiusSqr then
            entity:destroy()
        elseif entityPos.z ~= 0 then
            local rigidBody = getComponent(entity, RigidBodyComponent)
            rigidBody.dynamicProperties.position.z = 0
            rigidBody.dynamicProperties:touch()
        end
    end

    
    --Spawn entities
    for id,spawnType in pairs(self.spawnTypes) do
        --To actually spawn a given entity for a given attempt, two conditions should be met.
        --The first condition is a random chance that adjusts the spawn frequency to the approprate
        --amount. The second condition is whether the entity will spawn in a valid position.
        --It is checked when the first condition is met and a position
        --for the entity has been decided.
        
        --To allow more than one entity of each type to spawn per spawn cycle, the SpawnSystem
        --attempts to spawn each given entity multiple times depending on the spawnFrequency.
        --numAttempts stores how many times the SpawnSystem attempts to spawn the given entity.
        local numAttempts = math.max(1, math.ceil(spawnType.spawnFrequency * 2))
        for i = 1, numAttempts do
            if rng:getReal(0,1) < spawnType.spawnFrequency / numAttempts then
                --First condition passed. Choose a location for the entity.
                
                --A random location in the square of sidelength 2*spawnRadius
                --centered on the player is chosen. The corners
                --of the square are outside the spawning region, but they
                --will fail the second condition, so entities still only
                --spawn within the spawning region.
                local xDist = rng:getReal(-1,1) * spawnType.spawnRadius
                local yDist = rng:getReal(-1,1) * spawnType.spawnRadius
                local zDist = 0
                
                --Distance from the player.
                local displacement = Vector3(xDist, yDist, zDist)
                local distSqr = displacement:squaredLength()
                
                --Distance from the location of the player in the previous spawn cycle.
                local displacementPrev = displacement + playerPos - self.playerPosPrev
                local distSqrPrev = displacementPrev:squaredLength()
                
                if distSqr <= spawnType.spawnRadiusSqr and distSqrPrev > spawnType.spawnRadiusSqr then
                    --Second condition passed. Spawn the entity.
                    local entity = spawnType.factoryFunction(playerPos + displacement)
                    if entity then
                        local spawnComponent = SpawnedComponent.new()
                        spawnComponent.spawnRadiusSqr = spawnType.spawnRadiusSqr
                        entity:addComponent(spawnComponent)
                    end
                end
            end
        end
    end
    
    --Update previous player location.
    self.playerPosPrev.x = playerNode.transform.position.x
    self.playerPosPrev.y = playerNode.transform.position.y
    self.playerPosPrev.z = playerNode.transform.position.z
end

-- Override from System
function SpawnSystem:update(renderTime, logicTime)
    self.timeSinceLastCycle = self.timeSinceLastCycle + logicTime
    
    --Perform spawn cycle if necessary (Reason for "if" rather than "while" stated below)
    if self.timeSinceLastCycle > SPAWN_INTERVAL then        
        self:_doSpawnCycle()
        
        --Spawn interval does not affect spawning logic. Therefore, at most one spawn
        --cycle will be done per frame.
         self.timeSinceLastCycle = 0
    end
end

File: src/microbe_stage/spawn_system.cpp
#include "microbe_stage/spawn_system.h"

// #include <algorithm>
// #include <cmath>
// #include <OgreVector3.h>
// #include <unordered_map>

// #include "scripting/luajit.h"
// #include "engine/component_factory.h"
// #include "engine/entity_filter.h"
// #include "engine/game_state.h"
// #include "engine/engine.h"
// #include "engine/entity.h"
#include "engine/player_data.h"
// #include "engine/rng.h"
// #include "engine/serialization.h"
// #include "engine/typedefs.h"
// #include "ogre/scene_node_system.h"

#include "generated/cell_stage_world.h"

#include "ThriveGame.h"

#include <Utility/Random.h>

using namespace thrive;

////////////////////////////////////////////////////////////////////////////////
// SpawnedComponent
////////////////////////////////////////////////////////////////////////////////

SpawnedComponent::SpawnedComponent(double newSpawnRadiusSqr) :
    Leviathan::Component(TYPE), spawnRadiusSqr(newSpawnRadiusSqr)
{}

// void
// SpawnedComponent::load(
//     const StorageContainer& storage
// ) {
//     Component::load(storage);
//     spawnRadiusSqr = storage.get<double>("spawnRadiusSqr");
// }


// StorageContainer
// SpawnedComponent::storage() const {
//     StorageContainer storage = Component::storage();
//     storage.set<double>("spawnRadiusSqr", spawnRadiusSqr);
//     return storage;
// }

////////////////////////////////////////////////////////////////////////////////
// SpawnSystem
////////////////////////////////////////////////////////////////////////////////
struct SpawnSystem::Implementation {
    SpawnerTypeId nextId = 0;
    std::unordered_map<SpawnerTypeId, SpawnType> spawnTypes;
    Float3 previousPlayerPosition = Float3(0, 0, 0);
    float timeSinceLastUpdate = 0;
};

// void SpawnSystem::luaBindings(
//     sol::state &lua
// ){
//     lua.new_usertype<SpawnSystem>("SpawnSystem",

//         sol::constructors<sol::types<>>(),

//         sol::base_classes, sol::bases<System>(),

//         "init", &SpawnSystem::init,

//         "addSpawnType", &SpawnSystem::addSpawnType,

//         "removeSpawnType", &SpawnSystem::removeSpawnType
//     );
// }

SpawnerTypeId
    SpawnSystem::addSpawnType(
        std::function<ObjectID(CellStageWorld&, Float3)> factoryFunction,
        double spawnDensity,
        double spawnRadius)
{
    SpawnType newSpawnType;
    newSpawnType.factoryFunction = factoryFunction;
    newSpawnType.spawnRadius = spawnRadius;
    newSpawnType.spawnRadiusSqr = std::pow(spawnRadius, 2);
    newSpawnType.spawnFrequency =
        spawnDensity * newSpawnType.spawnRadiusSqr * 4;
    newSpawnType.id = m_impl->nextId;
    m_impl->nextId++;
    m_impl->spawnTypes[newSpawnType.id] = newSpawnType;
    return newSpawnType.id;
}

void
    SpawnSystem::removeSpawnType(SpawnerTypeId spawnId)
{
    m_impl->spawnTypes.erase(spawnId);
}

bool
    SpawnSystem::updateDensity(SpawnerTypeId spawnId, double spawnDensity)
{
    const auto found = m_impl->spawnTypes.find(spawnId);

    if(found == m_impl->spawnTypes.end())
        return false;

    found->second.spawnFrequency =
        spawnDensity * found->second.spawnRadiusSqr * 4;
    return true;
}

void
    SpawnSystem::Release()
{
    Clear();
}

void
    SpawnSystem::Clear()
{
    Leviathan::System<
        std::tuple<SpawnedComponent&, Leviathan::Position&>>::Clear();
    LOG_INFO("Clearing spawn system spawners");
    m_impl->spawnTypes.clear();
    m_impl->previousPlayerPosition = Float3(0, 0, 0);
    m_impl->timeSinceLastUpdate = 0;
}
// ------------------------------------ //

SpawnSystem::SpawnSystem() : m_impl(new Implementation()) {}

SpawnSystem::~SpawnSystem() {}

void
    SpawnSystem::Run(CellStageWorld& world, float elapsed)
{
    if(!world.GetNetworkSettings().IsAuthoritative)
        return;

    m_impl->timeSinceLastUpdate += elapsed;

    while(m_impl->timeSinceLastUpdate > SPAWN_INTERVAL) {
        m_impl->timeSinceLastUpdate -= SPAWN_INTERVAL;

        Float3 playerPosition = Float3(0, 0, 0);

        // Hybrid client-server version
        if(ThriveGame::Get()) {

            // Getting the player position.

            auto controlledEntity =
                ThriveGame::Get()->playerData().activeCreature();

            // Skip if no player entity //
            if(controlledEntity == NULL_OBJECT)
                continue;

            try {

                playerPosition = world.GetComponent_Position(controlledEntity)
                                     .Members._Position;

            } catch(const Leviathan::NotFound& e) {

                LOG_WARNING(
                    "SpawnSystem: no Position component in activeCreature, "
                    "exception:");
                e.PrintToLog();
                return;
            }
        }

        // Remove the y-position from player position
        playerPosition.Y = 0;
        int entitiesDeleted = 0;

        // Despawn entities.
        for(const auto& entry : CachedComponents.GetIndex()) {
            // delete a max of two entities per step to reduce lag from deleting
            // tons of entities at once
            if(entitiesDeleted < 2) {
                SpawnedComponent& spawnedComponent = std::get<0>(*entry.second);
                const Float3 spawnedEntityPosition =
                    std::get<1>(*entry.second).Members._Position;
                float squaredDistance =
                    (playerPosition - spawnedEntityPosition).LengthSquared();
                // If the entity is too far away from the player, despawn it.
                if(squaredDistance > spawnedComponent.spawnRadiusSqr) {
                    entitiesDeleted++;
                    world.QueueDestroyEntity(entry.first);
                }
            } else {
                // get out of loop if you hit max
                break;
            }
        }

        Leviathan::Random* random = Leviathan::Random::Get();

        // Spawn new entities.
        for(auto& st : m_impl->spawnTypes) {
            /*
            To actually spawn a given entity for a given attempt, two
            conditions should be met.  The first condition is a random
            chance that adjusts the spawn frequency to the approprate
            amount. The second condition is whether the entity will
            spawn in a valid position.  It is checked when the first
            condition is met and a position for the entity has been
            decided.

            To allow more than one entity of each type to spawn per
            spawn cycle, the SpawnSystem attempts to spawn each given
            entity multiple times depending on the spawnFrequency.
            numAttempts stores how many times the SpawnSystem attempts
            to spawn the given entity.
            */
            SpawnType& spawnType = st.second;
            unsigned numAttempts =
                std::max(int(spawnType.spawnFrequency * 2), 1);
            for(unsigned i = 0; i < numAttempts; i++) {
                if(random->GetNumber(0.0f, numAttempts) <
                    spawnType.spawnFrequency) {
                    /*
                    First condition passed. Choose a location for the entity.

                    A random location in the square of sidelength 2*spawnRadius
                    centered on the player is chosen. The corners
                    of the square are outside the spawning region, but they
                    will fail the second condition, so entities still only
                    spawn within the spawning region.
                    */
                    float distanceX = random->GetNumber(
                        static_cast<float>(-spawnType.spawnRadius),
                        spawnType.spawnRadius);
                    float distanceZ = random->GetNumber(
                        static_cast<float>(-spawnType.spawnRadius),
                        spawnType.spawnRadius);

                    // Distance from the player.
                    Float3 displacement(distanceX, 0, distanceZ);
                    float squaredDistance = displacement.LengthSquared();

                    // Distance from the location of the player in the previous
                    // spawn cycle.
                    Float3 previousDisplacement =
                        displacement + playerPosition -
                        m_impl->previousPlayerPosition;
                    float previousSquaredDistance =
                        previousDisplacement.LengthSquared();

                    if(squaredDistance <= spawnType.spawnRadiusSqr &&
                        previousSquaredDistance > spawnType.spawnRadiusSqr) {
                        // Second condition passed. Spawn the entity.
                        ObjectID spawnedEntity = spawnType.factoryFunction(
                            world, playerPosition + displacement);

                        // Giving the new entity a spawn component.
                        if(spawnedEntity != NULL_OBJECT) {

                            try {
                                world.Create_SpawnedComponent(
                                    spawnedEntity, spawnType.spawnRadiusSqr);
                            } catch(const Leviathan::Exception& e) {

                                LOG_ERROR("SpawnSystem failed to add "
                                          "SpawnedComponent, "
                                          "exception:");
                                e.PrintToLog();
                            }
                        }
                    }
                }
            }
        }

        // Updating the previous player location.
        m_impl->previousPlayerPosition = playerPosition;
    }
}
